# Problema 2

In [10]:
# Importamos las posibles librerias a utilizar en caso de requerir otra en el transcurso la agregamos
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [29]:
# Creamos un DataFrame a partir de los datos proporcionados en el ejercicion, agregando una columna de ticket para tenerlo como referencia
data = {'TicketNo': ['Ticket_1', 'Ticket_2', 'Ticket_3', 'Ticket_4', 'Ticket_5', 
                 'Ticket_6', 'Ticket_7', 'Ticket_8', 'Ticket_9', 'Ticket_10', 'Ticket_11'],
    'Products': [['bread', 'butter', 'wine', 'bananas', 'coffee', 'carrots'],
        ['tomatoes', 'onions', 'cheese', 'milk', 'potatoes'],
        ['beer', 'chips', 'asparagus', 'salsa', 'milk', 'apples'],
        ['olive oil', 'bread', 'butter', 'tomatoes', 'steak', 'carrots'],
        ['tomatoes', 'onions', 'chips', 'wine', 'ketchup', 'oranges'],
        ['bread', 'butter', 'beer', 'chips', 'milk'],
        ['butter', 'tomatoes', 'carrots', 'coffee', 'sugar'],
        ['tomatoes', 'onions', 'cheese', 'milk', 'potatoes'],
        ['bread', 'butter', 'ketchup', 'coffee', 'chicken wings'],
        ['butter', 'beer', 'chips', 'asparagus', 'apples'],
        ['tomatoes', 'onion', 'beer', 'chips', 'milk', 'coffee']]}
my_basket = pd.DataFrame(data)
my_basket

TicketNo                                           Products
0    Ticket_1    [bread, butter, wine, bananas, coffee, carrots]
1    Ticket_2         [tomatoes, onions, cheese, milk, potatoes]
2    Ticket_3      [beer, chips, asparagus, salsa, milk, apples]
3    Ticket_4  [olive oil, bread, butter, tomatoes, steak, ca...
4    Ticket_5  [tomatoes, onions, chips, wine, ketchup, oranges]
5    Ticket_6                 [bread, butter, beer, chips, milk]
6    Ticket_7         [butter, tomatoes, carrots, coffee, sugar]
7    Ticket_8         [tomatoes, onions, cheese, milk, potatoes]
8    Ticket_9    [bread, butter, ketchup, coffee, chicken wings]
9   Ticket_10           [butter, beer, chips, asparagus, apples]
10  Ticket_11       [tomatoes, onion, beer, chips, milk, coffee]

In [32]:
# Ahora, vamos a crear una copia para expandir los productos 
expanded = my_basket.explode('Products')
# Consolidar por TicketNo y producto (contando ocurrencias)
basket = (expanded.groupby(['TicketNo', 'Products']).size().unstack(fill_value=0).reset_index().set_index('TicketNo'))
basket

Products   apples  asparagus  bananas  beer  bread  butter  carrots  cheese  \
TicketNo                                                                      
Ticket_1        0          0        1     0      1       1        1       0   
Ticket_10       1          1        0     1      0       1        0       0   
Ticket_11       0          0        0     1      0       0        0       0   
Ticket_2        0          0        0     0      0       0        0       1   
Ticket_3        1          1        0     1      0       0        0       0   
Ticket_4        0          0        0     0      1       1        1       0   
Ticket_5        0          0        0     0      0       0        0       0   
Ticket_6        0          0        0     1      1       1        0       0   
Ticket_7        0          0        0     0      0       1        1       0   
Ticket_8        0          0        0     0      0       0        0       1   
Ticket_9        0          0        0     0      1       1        0       0   

Products   chicken wings  chips  ...  olive oil  onion  onions  oranges  \
TicketNo                         ...                                      
Ticket_1               0      0  ...          0      0       0        0   
Ticket_10              0      1  ...          0      0       0        0   
Ticket_11              0      1  ...          0      1       0        0   
Ticket_2               0      0  ...          0      0       1        0   
Ticket_3               0      1  ...          0      0       0        0   
Ticket_4               0      0  ...          1      0       0        0   
Ticket_5               0      1  ...          0      0       1        1   
Ticket_6               0      1  ...          0      0       0        0   
Ticket_7               0      0  ...          0      0       0        0   
Ticket_8               0      0  ...          0      0       1        0   
Ticket_9               1      0  ...          0      0       0        0   

Products   potatoes  salsa  steak  sugar  tomatoes  wine  
TicketNo                                                  
Ticket_1          0      0      0      0         0     1  
Ticket_10         0      0      0      0         0     0  
Ticket_11         0      0      0      0         1     0  
Ticket_2          1      0      0      0         1     0  
Ticket_3          0      1      0      0         0     0  
Ticket_4          0      0      1      0         1     0  
Ticket_5          0      0      0      0         1     1  
Ticket_6          0      0      0      0         0     0  
Ticket_7          0      0      0      1         1     0  
Ticket_8          1      0      0      0         1     0  
Ticket_9          0      0      0      0         0     0  

[11 rows x 23 columns]

In [33]:
# Aplicamos el algoritmo "a priori" para localizar los artículos frecuentemente comprados en la base de datos
from mlxtend.frequent_patterns import apriori

In [48]:
# analisis "a priori" con un soporte mínimo del 25% tomo este criterio porque nuestra basde de datos en muy pequeña
articulos_frecuentes = apriori(basket, min_support = 0.25, use_colnames = True).sort_values('support', ascending = False).reset_index(drop = True)
articulos_frecuentes['length'] = articulos_frecuentes['itemsets'].apply(lambda x:len(x))
articulos_frecuentes

/home/9e16b0a4-46cf-4ed8-a80e-6fb1cbba99a2/.local/lib/python3.11/site-packages/mlxtend/frequent_patterns/fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support             itemsets  length
0   0.545455           (tomatoes)       1
1   0.545455             (butter)       1
2   0.454545              (chips)       1
3   0.454545               (milk)       1
4   0.363636               (beer)       1
5   0.363636      (butter, bread)       2
6   0.363636              (bread)       1
7   0.363636        (chips, beer)       2
8   0.363636             (coffee)       1
9   0.272727             (onions)       1
10  0.272727         (milk, beer)       2
11  0.272727            (carrots)       1
12  0.272727    (butter, carrots)       2
13  0.272727     (butter, coffee)       2
14  0.272727        (chips, milk)       2
15  0.272727     (milk, tomatoes)       2
16  0.272727   (onions, tomatoes)       2
17  0.272727  (chips, milk, beer)       3

Vemos los articulos y el soporte en función de la frecuencia que llegamos a encontrar, eEsto es preeliminar tenemos un total de82 diferentes articulos
• Ahora vamos a hacer las reglas de asociación:

In [49]:
 ## Definimos las reglas de asociación
association_rules(articulos_frecuentes, metric = 'confidence',
                 min_threshold = 0.50).sort_values('confidence', ascending = False).reset_index(drop=True)

antecedents    consequents  antecedent support  consequent support  \
0         (bread)       (butter)            0.363636            0.545455   
1          (beer)        (chips)            0.363636            0.454545   
2    (milk, beer)        (chips)            0.272727            0.454545   
3       (carrots)       (butter)            0.272727            0.545455   
4   (chips, milk)         (beer)            0.272727            0.363636   
5        (onions)     (tomatoes)            0.272727            0.545455   
6         (chips)         (beer)            0.454545            0.363636   
7          (beer)  (chips, milk)            0.363636            0.272727   
8          (beer)         (milk)            0.363636            0.454545   
9   (chips, beer)         (milk)            0.363636            0.454545   
10       (coffee)       (butter)            0.363636            0.545455   
11       (butter)        (bread)            0.545455            0.363636   
12         (milk)     (tomatoes)            0.454545            0.545455   
13        (chips)         (milk)            0.454545            0.454545   
14         (milk)         (beer)            0.454545            0.363636   
15        (chips)   (milk, beer)            0.454545            0.272727   
16         (milk)  (chips, beer)            0.454545            0.363636   
17         (milk)        (chips)            0.454545            0.454545   
18     (tomatoes)         (milk)            0.545455            0.454545   
19     (tomatoes)       (onions)            0.545455            0.272727   
20       (butter)       (coffee)            0.545455            0.363636   
21       (butter)      (carrots)            0.545455            0.272727   

     support  confidence      lift  representativity  leverage  conviction  \
0   0.363636    1.000000  1.833333               1.0  0.165289         inf   
1   0.363636    1.000000  2.200000               1.0  0.198347         inf   
2   0.272727    1.000000  2.200000               1.0  0.148760         inf   
3   0.272727    1.000000  1.833333               1.0  0.123967         inf   
4   0.272727    1.000000  2.750000               1.0  0.173554         inf   
5   0.272727    1.000000  1.833333               1.0  0.123967         inf   
6   0.363636    0.800000  2.200000               1.0  0.198347    3.181818   
7   0.272727    0.750000  2.750000               1.0  0.173554    2.909091   
8   0.272727    0.750000  1.650000               1.0  0.107438    2.181818   
9   0.272727    0.750000  1.650000               1.0  0.107438    2.181818   
10  0.272727    0.750000  1.375000               1.0  0.074380    1.818182   
11  0.363636    0.666667  1.833333               1.0  0.165289    1.909091   
12  0.272727    0.600000  1.100000               1.0  0.024793    1.136364   
13  0.272727    0.600000  1.320000               1.0  0.066116    1.363636   
14  0.272727    0.600000  1.650000               1.0  0.107438    1.590909   
15  0.272727    0.600000  2.200000               1.0  0.148760    1.818182   
16  0.272727    0.600000  1.650000               1.0  0.107438    1.590909   
17  0.272727    0.600000  1.320000               1.0  0.066116    1.363636   
18  0.272727    0.500000  1.100000               1.0  0.024793    1.090909   
19  0.272727    0.500000  1.833333               1.0  0.123967    1.454545   
20  0.272727    0.500000  1.375000               1.0  0.074380    1.272727   
21  0.272727    0.500000  1.833333               1.0  0.123967    1.454545   

    zhangs_metric   jaccard  certainty  kulczynski  
0        0.714286  0.666667   1.000000    0.833333  
1        0.857143  0.800000   1.000000    0.900000  
2        0.750000  0.600000   1.000000    0.800000  
3        0.625000  0.500000   1.000000    0.750000  
4        0.875000  0.750000   1.000000    0.875000  
5        0.625000  0.500000   1.000000    0.750000  
6        1.000000  0.800000   0.685714    0.900000  
7        1.000000  0.750000   0.656250

In [52]:
# Definimos las reglas de asociación
association_rules(articulos_frecuentes, metric = 'lift',
                 min_threshold = 1.01).sort_values('lift', ascending = False).reset_index(drop=True)

antecedents    consequents  antecedent support  consequent support  \
0          (beer)  (chips, milk)            0.363636            0.272727   
1   (chips, milk)         (beer)            0.272727            0.363636   
2         (chips)         (beer)            0.454545            0.363636   
3          (beer)        (chips)            0.363636            0.454545   
4         (chips)   (milk, beer)            0.454545            0.272727   
5    (milk, beer)        (chips)            0.272727            0.454545   
6         (bread)       (butter)            0.363636            0.545455   
7      (tomatoes)       (onions)            0.545455            0.272727   
8        (onions)     (tomatoes)            0.272727            0.545455   
9        (butter)        (bread)            0.545455            0.363636   
10      (carrots)       (butter)            0.272727            0.545455   
11       (butter)      (carrots)            0.545455            0.272727   
12  (chips, beer)         (milk)            0.363636            0.454545   
13         (beer)         (milk)            0.363636            0.454545   
14         (milk)         (beer)            0.454545            0.363636   
15         (milk)  (chips, beer)            0.454545            0.363636   
16       (coffee)       (butter)            0.363636            0.545455   
17       (butter)       (coffee)            0.545455            0.363636   
18        (chips)         (milk)            0.454545            0.454545   
19         (milk)        (chips)            0.454545            0.454545   
20         (milk)     (tomatoes)            0.454545            0.545455   
21     (tomatoes)         (milk)            0.545455            0.454545   

     support  confidence      lift  representativity  leverage  conviction  \
0   0.272727    0.750000  2.750000               1.0  0.173554    2.909091   
1   0.272727    1.000000  2.750000               1.0  0.173554         inf   
2   0.363636    0.800000  2.200000               1.0  0.198347    3.181818   
3   0.363636    1.000000  2.200000               1.0  0.198347         inf   
4   0.272727    0.600000  2.200000               1.0  0.148760    1.818182   
5   0.272727    1.000000  2.200000               1.0  0.148760         inf   
6   0.363636    1.000000  1.833333               1.0  0.165289         inf   
7   0.272727    0.500000  1.833333               1.0  0.123967    1.454545   
8   0.272727    1.000000  1.833333               1.0  0.123967         inf   
9   0.363636    0.666667  1.833333               1.0  0.165289    1.909091   
10  0.272727    1.000000  1.833333               1.0  0.123967         inf   
11  0.272727    0.500000  1.833333               1.0  0.123967    1.454545   
12  0.272727    0.750000  1.650000               1.0  0.107438    2.181818   
13  0.272727    0.750000  1.650000               1.0  0.107438    2.181818   
14  0.272727    0.600000  1.650000               1.0  0.107438    1.590909   
15  0.272727    0.600000  1.650000               1.0  0.107438    1.590909   
16  0.272727    0.750000  1.375000               1.0  0.074380    1.818182   
17  0.272727    0.500000  1.375000               1.0  0.074380    1.272727   
18  0.272727    0.600000  1.320000               1.0  0.066116    1.363636   
19  0.272727    0.600000  1.320000               1.0  0.066116    1.363636   
20  0.272727    0.600000  1.100000               1.0  0.024793    1.136364   
21  0.272727    0.500000  1.100000               1.0  0.024793    1.090909   

    zhangs_metric   jaccard  certainty  kulczynski  
0        1.000000  0.750000   0.656250    0.875000  
1        0.875000  0.750000   1.000000    0.875000  
2        1.000000  0.800000   0.685714    0.900000  
3        0.857143  0.800000   1.000000    0.900000  
4        1.000000  0.600000   0.450000    0.800000  
5        0.750000  0.600000   1.000000    0.800000  
6        0.714286  0.666667   1.000000    0.833333  
7        1.000000  0.500000   0.312500

## Resumen del Análisis de las Reglas de Asociación (confidence).
Para este ejercicio las reglas las filtré con un umbral de confianza (`min_threshold`) de 0.5, lo que nos garantiza que solo se consideren relaciones significativas entre productos.
#### **Principales Patrones de Consumo**
Los patrones en las reglas de los clientes son los siguientes:

- **Combinaciones Fuertemente Relacionadas**:
- Los clientes que compran `bread` siempre compran `butter`
- Los clientes que compran `beer` tienen una alta probabilidad de comprar `chips`esto en el  75% de los casos
- Los clientes que compran `carrots` tienden a comprar `butter` en el **80%** de los casos
- La combinación de `chips` y `milk` está fuertemente asociada con la compra de `beer` igual en un 75% de las veces

- **Productos Individuales Populares**:
Por otro lado tenemos productos individuales que son muy populares, como por ejemplo: 
- **tomatoes**: Aparece como consecuencia en varias reglas (por ejemplo, `onions -> tomatoes`).   
- **milk**: También es un producto popular, apareciendo tanto como antecedente como consecuente en varias reglas. Su alta frecuencia indica que es un artículo esencial en muchas transacciones.
  
#### **Oportunidades de Marketing**
Basándose en estas reglas, se pueden implementar estrategias de marketing efectivas, como lo son las **Promociones Cruzadas**, **Colocación Estratégica**, **Recomendaciones Personalizadas** e incluso la **Segmentación de Clientes**


### Resumen del Análisis de las Reglas de Asociación (lift)

Para este caso, las reglas han sido filtradas con un umbral de **lift** (`min_threshold`) de 1.01, lo que garantiza que solo se consideren relaciones significativas entre productos.

#### **Principales Patrones de Consumo**
Los patrones que se observan son los siguientes:

- **Combinaciones Fuertemente Relacionadas (Altos Valores de Lift)**:
  
- Los clientes que compran `beer` tienden a comprar `chips` y `milk` juntos, lo que sugiere una relación fuerte entre estos productos.
- La combinación de `chips` y `milk` está altamente asociada con la compra de `beer`.
- La relación entre `tomatoes` y `onions` es significativa, ya que la compra de uno incrementa considerablemente la probabilidad de comprar el otro.
- Los clientes que compran `butter` tienen una alta probabilidad de comprar `carrots`.

- **Productos Individuales Populares**:
- **`tomatoes` y `onions`**: Ambos productos aparecen en varias reglas, mostrando una relación estrecha. Por ejemplo, `onions -> tomatoes` tiene un **lift de 1.45**, lo que indica que la compra de `onions` incrementa significativamente la probabilidad de comprar `tomatoes`.
- **`butter` y `carrots`**: La regla `butter -> carrots` tiene un **lift de 1.83**, lo que sugiere que los clientes que compran `butter` tienden a comprar `carrots` con mayor frecuencia de lo esperado por casualidad.

#### **Oportunidades de Marketing**
Al igual que en el otro método, estos resultados son practicamente iguales, por lo tanto con más confienza podemos implementar estrategias de marketing ya mencionadas anteriormente.